pkg/kubelet/container/runtime.go

```diff
@@ -333,6 +333,8 @@ type Status struct {
 	// Message written by the container before exiting (stored in
 	// TerminationMessagePath).
 	Message string
+	// CPU and memory resources for this container
+	Resources v1.ResourceRequirements
```

Comment:
Do you mean we store runtimeapi.ContainerResources? Or do we create a new type to hold values we care about?

Reply:
This is a big change to do. This will delay us into 1.23.

Reply:
This is not as critical. If time is sensitive, it is OK to do a followup cleanup

Reply:
The reason it is v1.ResourceRequirements is because the core operating logic also uses the values in kubeContainerStatus.Resources in [computePodActions](https://github.com/vinaykul/kubernetes/blob/restart-free-pod-vertical-scaling-core/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L716) and [updatePodContainerResources](https://github.com/vinaykul/kubernetes/blob/restart-free-pod-vertical-scaling-core/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L1110) which was quite painstakingly put-together and tested for various edge cases. Generating it in apiPodStatus conversion is the easy part.

So changing this involves potentially resetting the testing of this core logic and rewriting unit tests. But it may not be as bad as I initially thought if I can invoke the conversion helper to return ResourceLists similar to what's being currently done in toKubeContainerStatus, and plug in the ResourceLists where needed.

The perf impact of this should not be too bad because this is not really hot-path. It is invoked as long as resize is pending, and I don't imagine VPA will constantly keep resizing pods - it should be a fairly uncommon even.

We can definitely do this in Beta. I'll add a TODO for now and see if I can pull this in (if there's not much changes requested from core & api review). That way I don't even have to deal with it later.

Reply:
> The reason it is v1.ResourceRequirements is because the core operating logic also uses the values in kubeContainerStatus.Resources in computePodActions and updatePodContainerResources which was quite painstakingly put-together and tested for various edge cases. Generating it in apiPodStatus conversion is the easy part.

Actually, because of this I feel more strongly that we should use the internal representation.

The reason is what we discussed about the information loss from api resource limit / request to the actual cgroup settings. Within kuberuntime manager, at the end, we should use the actual cgroup settings to discover the diff, and decide what actions to take, instead of relying on the the api resource limit / request.

The api resource limit / request should only be used when generating the actual desired cgroup settings, and generating the actual api status, that would give us more room to do adhoc conversion for the api status to avoid user confusion as well, because it won't affect the internal logic.

Anyway, if the current logic is correct, we can revisit this in beta, but I think having an internal representation seems more useful after you mentioned that. :)

Reply:
I've taken a TODO for now, but this makes a lot of sense.